lib/Ast.ml

```diff
@@ -1,223 +0,0 @@
-(** Abstract syntax. *)
-open Core_kernel
-
-(* for auto generating s-exp *)
-
-(* == Source code locations for semantic errors == *)
-type location =
-  | Location of Lexing.position * Lexing.position  (** delimited location *)
-  | Nowhere  (** no location *)
-
-(* == Unsized types == *)
-type originblock =
-  | Primitives
-  | Functions
-  | Data
-  | TData
-  | Param
-  | TParam
-  | Model
-  | GQuant
-
-and unsizedtype =
-  | Int
-  | Real
-  | Vector
-  | RowVector
-  | Matrix
-  | Array of unsizedtype
-  | Fun of (originblock * unsizedtype) list * returntype
-  | PrimitiveFunction
-
-and returntype = Void | ReturnType of unsizedtype
-
-(* == Expressions == *)
-and identifier = {name: string; id_loc: location sexp_opaque [@compare.ignore]}
-
-and infixop =
-  | Plus
-  | Minus
-  | Times
-  | Divide
-  | Modulo
-  | LDivide
-  | EltTimes
-  | EltDivide
-  | Exp
-  | Or
-  | And
-  | Equals
-  | NEquals
-  | Less
-  | Leq
-  | Greater
-  | Geq
-
-and prefixop = Not | UMinus | UPlus
-
-and postfixop = Transpose
-
-and 'e index =
-  | All
-  | Single of 'e
-  | Upfrom of 'e
-  | Downfrom of 'e
-  | Between of 'e * 'e
-  | Multiple of 'e
-
-and 'e expression =
```

Comment:
We talked briefly about how if, in the body here, we just wrote `'e expression` everywhere we wrote `'t` or something then `'t` becomes the type of the metadata. this might be easier to understand and only a little bit more annoying to write.

Reply:
Could you make the proposal more precise? The version I can imagine would really be quite verbose, but I could well be missing something.